### ISSUE TITLE
Multi Loss Function

### ISSUE THREAD

**Issue by slegall56** [fchollet/keras#4126]
Hi, 
I want to implement a neural network with two loss function as in [http://ydwen.github.io/papers/WenECCV16.pdf]

How can I implement this in Keras? 

Do I have to split my last dense layer and call a loss function layer for each output? 
Or do I have to create my own custom layer which has to compute both softmax and center loss ? 

Regards

Sebastien

**Comment by slegall56** [comment]
I have an idea

I started to create a network with 2 outputs such as :
```

inputs = Input(shape=(100,100,3))
...
fc = Dense(100)(#previousLayer#)
softmax = Softmax(fc)
model = Model(input, output=[softmax, fc])
model.compile(optimizer='sgd', 
              loss=['categorical_crossentropy', 'center_loss'],
              metrics=['accuracy'], loss_weights=[1., 0.2])
```
First of all, doing like this, is it the good way to proceed?

Secondly, I don't know how to implement the center_loss in keras. Center_loss looks like mean square error but instead of comparing values to fixed labels, it compares values to data updated at each iteration.

Thank you for your help

**Comment by dieuwkehupkes** [comment]
Maybe you could also - instead of creating two output layers - create a custom loss-function. This would also allow you to put the hyperparameter they mention in the paper that balances the two supervision signals.

I am not sure if I understand correctly how this center loss works. At every iteration you update the mean of the predicted class? or of the class it should have been? Maybe you could somehow integrate this into your architecture? If you have some fixed weight layers that compute the class centers, you could instead of using the class center as prediction create a 'subtraction layer' that computes the difference between the layer that computes the right classweight and the ouput of the network and use mse to train that (with target 0). Then you cannot use the custom lossfunction anymore, but you could still implement the hyperparameter by the fixed weights in this subtraction layer. If this doesn't make any sense, could you maybe tell me how you would compute the Y_pred in each iteration if you could just input them as you like?

**Comment by slegall56** [comment]
Well, for me creating one custom loss-function doing both of the loss is equal to having two different loss functions and the hyperparameters is set by the loss_weights parameters in compile function. 

So to simplyfy the model, let say I want just to apply the center loss function only. I need to have a code like this : 


So I need to create a loss function which stores centers and update them at each epoch

**Comment by belavenir** [comment]
@slegall56 I'm doing the same thing right now. Your initial idea is a good try. Rather than caffe, keras has no argument " loss_weights", so maybe it asks for a self-defined loss function. I think a model with two outputs is ok. 
`model.compile(optimizer='sgd', 
              loss=['categorical_crossentropy', 'center_loss'],
              metrics=['accuracy'], loss_weights=[1., 0.2])`

**Comment by slegall56** [comment]
Well, indeed 2 outputs with loss_weights parameters seems to work. We just need to define our own center loss function in objectives.py

**Comment by biprajiman** [comment]
I recently tried something similar. Give this a shot. I have not verified it but seems to be working. Might as well be a verification.
```python
def prepare_model():
      inputs = Input(shape=(100,100,3))
      ...
      fc = Dense(100)(#previousLayer#)
      softmax = Softmax(fc)

       def custom_loss(y_true, y_pred):
             loss1=softmax(y_true,y_pred)
             loss2=center_loss(y_true,fc)
             return loss1+lambda*loss2

        model = Model(input, output=softmax)
        model.compile(optimizer='sgd',
              loss=custom_loss,
              metrics=['accuracy'])
       return model
```

**Comment by wangxianliang** [comment]
Modified from https://github.com/davidsandberg/facenet/blob/master/src/facenet.py
```python
class CenterLossLayer(Layer):
    """Center loss based on the paper "A Discriminative Feature Learning Approach for Deep Face Recognition"
       (http://ydwen.github.io/papers/WenECCV16.pdf)
    """
    def __init__(self, alpha, num_classes, **kwargs):        
        self._alpha = alpha
        self._num_classes = num_classes
        super(CenterLossLayer, self).__init__(**kwargs)

    def build(self, input_shape):
        # Create class center variable for this layer.        
        self._centers = self.add_weight(name='centers',
                                        shape=(self._num_classes, input_shape[1]),
                                        initializer='zeros',
                                        trainable=False)
        super(CenterLossLayer, self).build(input_shape)  # Be sure to call this somewhere!

    def call(self, y_true, y_pred):
        labels = K.reshape(y_true, [-1])
        centers_batch = K.gather(self._centers, labels)
        diff = (1 - self._alpha) * (centers_batch - y_pred)
        self._centers = K.scatter_sub(self._centers, labels, diff)
        loss = K.reduce_mean(K.square(y_pred - centers_batch))
        return loss

    def compute_output_shape(self, input_shape):
        return (1,)
```

**Comment by alyato** [comment]
@biprajiman how to understand the `loss2=center_loss(y_true,fc)`, do you explain it in detail.
Or give one snip for center_loss? Thanks.

**Comment by alyato** [comment]
@slegall56 Do you run it well?

**Comment by biprajiman** [comment]
@alyato if you have a look at the paper...it is a simple L2 norm square or you can use any other loss function based on your requirement.

**Comment by Tharun98** [comment]
Hii,
Suppose if i have 2 loss functions.Some of the weights are common for both the outputs. How will it update the weights using back propagation theorem? 
Thanks in advance.
